runtime/lua/vim/lsp.lua

```diff
@@ -16,24 +7,41 @@ local tbl_isempty, tbl_extend = vim.tbl_isempty, vim.tbl_extend
 local validate = vim.validate
 local if_nil = vim.F.if_nil
 
-local lsp = {
-  protocol = protocol,
-
-  handlers = default_handlers,
-
-  buf = require('vim.lsp.buf'),
-  diagnostic = require('vim.lsp.diagnostic'),
-  codelens = require('vim.lsp.codelens'),
-  inlay_hint = require('vim.lsp.inlay_hint'),
-  semantic_tokens = semantic_tokens,
-  util = util,
-
-  -- Allow raw RPC access.
-  rpc = lsp_rpc,
-
-  -- Export these directly from rpc.
-  rpc_response_error = lsp_rpc.rpc_response_error,
+local submodules = {
+  '_changetracking',
+  '_completion',
+  '_dynamic',
+  'buf',
+  'codelens',
+  'diagnostic',
+  'handlers',
+  'inlay_hint',
+  'log',
+  'protocol',
+  'rpc',
+  'semantic_tokens',
+  'tagfunc',
+  'util',
 }
+local dundarload_submodules = vim._dundarload_submodules('vim.lsp', submodules)
+local lsp = setmetatable({}, dundarload_submodules)
+
+lsp._changetracking = nil --- @module 'vim.lsp._changetracking'
+lsp._completion = nil --- @module 'vim.lsp._completion'
+lsp.util = nil --- @module 'vim.lsp.util'
+lsp.log = nil --- @module 'vim.lsp.log'
+lsp.protocol = nil --- @module 'vim.lsp.protocol'
+lsp.rpc = nil --- @module 'vim.lsp.rpc'
+lsp.diagnostic = nil --- @module 'vim.lsp.diagnostic'
```

Comment:
We need this for all the submodules as this type information will propagate to other files.